matrix_sdk/src/http_client.rs

```diff
@@ -112,12 +112,12 @@ impl HttpClient {
                     read_guard = session.read().await;
 
                     if let Some(session) = read_guard.as_ref() {
-                        Some(session.access_token.as_str())
+                        SendAccessToken::IfRequired(session.access_token.as_str())
```

Comment:
It should probably be configurable per request whether `IfRequired` or `Always` is used here, so the `Always` will be used for the profile requests that users were previously enabling `unstable-synapse-quirks` for.

Reply:
Then we might want to reintroduce the feature in the SDK and have it use `SendAccessToken::Always` for certain requests. Could also be done directly in that method by matching on the unique ruma request id?

Reply:
I don't think it needs to be behind a feature flag. matrix-sdk-common already unconditionally enabled ruma's compat feature, and this is very much a hack for increased compatibility. What I was imagining in terms of implementation was not as hacky as matching on the request ID but rather give this function a `force_auth: bool` parameter that decides whether `IfRequired` or `Always` is used. (or maybe a dedicated enum would be better for readability of the callsites)

Reply:
Sounds good. Does by chance a list exist which endpoints are affected?

Reply:
All the ones you changed in https://github.com/ruma/ruma/commit/37462b9ae16008cfc380a78e382c93aa46766c3d: `get_avatar_url`, `get_display_name`, `get_profile`. Not sure whether others are affected.

Reply:
Relevant part of synapse config: https://github.com/matrix-org/synapse/blob/ceaa76970fa0092bbdc35055c6f32dc63dd59960/docs/sample_config.yaml#L113

There don't seem to be other `require_auth` settings.

Reply:
> All the ones you changed in ruma/ruma@37462b9: get_avatar_url, get_display_name, get_profile. Not sure whether others are affected.

Right, I did that :sweat_smile: 

> Relevant part of synapse config

Interesting, thanks!

Reply:
@poljar We seem to have two possibilities here,

1.
  - Add an additional `force_auth`argument on `send_request` / `send`. That would require to refactor ~50 callsites and expose the argument to consumers
  - Add `force_auth` to `RequestConfig`. Would also require to refactor ~50 callsites and exposing to consumers would be optional since the argument could be crate-internal
2. Match on the request id string inside of `send_request`. This would have the downside that if `send` is used with a Custom event accidentally matching one of the id strings, it would trigger adding of authentication. The quirky ids could be mentioned in the methods docblock tho

Which one would you prefer?

Reply:
Couldn't we do 3. but not have the accidental match? Otherwise I think we'll have to do 2.

Reply:
No, but to quote myself from #ruma:matrix.org

> it seems exceedingly unlikely consumer would create custom endpoint definitions named the same as ones in Ruma

This is about the [name field in endpoint metadata](https://github.com/ruma/ruma/blob/12ec0fb1680ebc4fec4fbefbbd0890ae4eaf3a88/ruma-api/tests/ui/01-api-sanity-check.rs#L9). I think "Custom event" above was supposed to be "Custom endpoint".

Reply:
Ah so it's about that name field that's not part of the spec, not about matching the URL where the event might have a custom type, e.g. `/rooms/send/my.type` and then we accidentally match on the type.

I guess 3. is in that case fine as well, considering that a very small percentage of users will define custom endpoints at all. We should document somewhere the danger of doing so and picking a sensible name.

Reply:
Yeah, sorry for the confusion, I meant custom endpoints. However, I've realized that 2. would only require touching one client method actually, so I went with that and also exposed the configuration as it might actually be useful for consumers. Also introduced an `require_auth_for_profile_requests` feature to match the synapse config as mentioned by @jplatte (in this case it only forces auth for `get_avatar` as the other profile related things come from the store).